File: gfg/2024_6_june/c++/05_swapping_pairs_make_sum_equal.cpp
// APPROACH
// 1 calculate sum of each both array
// 2 if rem of sum diff is not equal to 0 then simply return -1
// 3 make two pointer i and j with value 0
// 4 sort both the arrays 
// 5 make target = diff_sum / 2 
// 6 rum while loop till size of both array with i and j pointer
// 7 if diff = (a[i] - b[j]) is equal to target return 1
// 8 diff < target i++
// 9 diff > target j++
// 10 return -1 out from loop 


// TIME COMPLEXITY
// O(N + M)

// SPACE COMPLEXITY
// O(1)

// CODE

class Solution {

  public:
    int findSwapValues(int a[], int n, int b[], int m) {
        
        int sum_a = 0;
        int sum_b = 0;
        
        for(int i = 0; i < n; i++) sum_a += a[i];
        for(int i = 0; i < m; i++) sum_b += b[i];
        
        int diff_sum = sum_a - sum_b;
        
        if(diff_sum % 2 != 0) return -1;
        
        int target = diff_sum / 2;
        
        int i = 0;
        int j = 0;
        
        sort(a, a+n);
        sort(b, b+m);
        
        while(i < n && j < m){
            int diff = a[i] - b[j];
            if(diff == target) return 1;
            else if(diff < target) i++;
            else j++;
        }
        return -1;
        
    }
};

File: gfg/2024_6_june/python/05_swapping_pairs_make_sum_equal.py
# INTUITION
# Let,
# Sum of array a = s1
# Sum of array b = s2

# Condition: s1 - a[i] + b[j] = s2 - b[j] + a[i]

# Simplification:
# 2 * (b[j] - a[i]) = s2 - s1
# b[j] - a[i] = (s2 - s1) / 2
# b[j] = a[i] + (s2 - s1) / 2
# Thus, we have to find above value in array a which equals to b[j]

# APPROACH
# 1. Find sum of both arrays, s1 and s2.
# 2. Evaluate target values as per above equation.
# 3. Sort the array b in order to apply binary search.
# 4. Traverse in array a and for each value, apply binary search and
#    find target value.
# 5. return -1 if target value not found for all element of array a.

# TIME COMPLEXITY
# O((n + m)*log(m)) 

# SPACE COMPLEXITY
# O(1)

# CODE

import bisect
class Solution:
    def findSwapValues(self, a, n, b, m):
        s1 = sum(a)
        s2 = sum(b)
        
        # if sum is odd, not possible
        if abs(s2 - s1) % 2:
            return -1
        
        sum_diff = (s2 - s1) // 2
        b.sort()
        
        for val in a:
            target = sum_diff + val
            if self.binary_search(b, target):
                return 1
        
        return -1
    
    def binary_search(self, arr, target):
        index = bisect.bisect_left(arr, target)
        return index < len(arr) and arr[index] == target